### PR TITLE
[WIP] Get external tensor shape

### PR DESCRIPTION
**Pull request by thinking-tower**

### CODE REVIEW

**Review by archibate — COMMENTED** [review]
The WIP code LGTM!
Remember to transform `ExternalTensorShapeAlongAxisExpression` into `ExternalTensorShapeAlongAxisStmt` in `taichi/transformers/lower_ast.cpp`.

**Comment by thinking-tower** [comment]
@archibate can you explain why I need to transform it in `taichi/transformers/lower_ast.cpp`? I'm doing it [here](https://github.com/thinking-tower/taichi/blob/feature/get-external-tensor-shape/taichi/ir/ir.cpp#L1055-L1060) atm.

```
void ExternalTensorShapeAlongAxisExpression::flatten(FlattenContext *ctx) {
  auto temp = ptr.cast<ExternalTensorExpression>();
  TI_ASSERT(0 <= axis && axis < temp->dim);
  ctx->push_back<ExternalTensorShapeAlongAxisStmt>(axis, temp->arg_id);
  stmt = ctx->back_stmt();
}
```

File: python/taichi/lang/expr.py [hunk 1]
```diff
@@ -144,6 +144,17 @@ def __hash__(self):
...
+        if self.ptr.is_external_var():
+            import taichi as ti
+            dim = ti.get_external_tensor_dim(self.ptr)
+            ret = [ti.get_external_tensor_shape_along_axis(self.ptr, i) for i in range(dim)]
+            class callable_tuple(tuple):
+                @deprecated('x.shape()', 'x.shape')
+                def __call__(self):
+                    return self
+
+            ret = callable_tuple(ret)
```
Comment (archibate) [inline python/taichi/lang/expr.py]:
```suggestion
```
The `x.shape()` API is to be deprecated soon anyway, breaking it in a newly-added feature is OK :>

Comment (thinking-tower) [reply python/taichi/lang/expr.py:156]: @archibate I still got `[<taichi_core.Expr object at 0x7f36e7c8d730>]` unfortunately.

File: taichi/transforms/ir_printer.cpp [hunk 2]
```diff
@@ -603,6 +603,10 @@ class IRPrinter : public IRVisitor {
     print("{}{} : stack acc adj {}, val = {}", stmt->type_hint(), stmt->name(),
           stmt->stack->name(), stmt->v->name());
   }
+
+  void visit(ExternalTensorShapeAlongAxisStmt *stmt) override {
+    print("axis {}, arg_id {}", stmt->axis, stmt->arg_id);
```
Comment (archibate) [inline taichi/transforms/ir_printer.cpp]:
```suggestion
    print("external_tensor_shape_alone_axis {}, arg_id {}", stmt->axis, stmt->arg_id);
```

File: taichi/backends/opengl/codegen_opengl.cpp [hunk 3]
```diff
@@ -693,6 +693,15 @@ class KernelGen : public IRVisitor {
     emit("{};", source);
   }
 
+  void visit(ExternalTensorShapeAlongAxisStmt *stmt) override {
+    used.buf_earg = true;
+    const auto name = stmt->short_name();
+    const auto arg_id = stmt->arg_id;
+    auto axis = stmt->axis;
+    emit("int {} = _earg_i32_[{} * {} + {}];", name, arg_id,
```
Comment (archibate) [inline taichi/backends/opengl/codegen_opengl.cpp:701]:
```suggestion
    used.buf_earg = true;
    emit("int {} = _earg_i32_[{} * {} + {}];", name, arg_id,
```

File: taichi/ir/frontend_ir.h [hunk 4]
```diff
@@ -585,4 +585,20 @@ class ConstExpression : public Expression {
   }
 };
 
+class ExternalTensorShapeAlongAxisExpression : public Expression {
+ public:
+  Expr ptr;
+  int axis;
+
+  std::string serialize() override {
+    return "ExternalTensorShapeAlongAxisExpression not implemented";
```
Comment (archibate) [inline taichi/ir/frontend_ir.h]:
```suggestion
    return fmt::format("external_tensor_shape_alpng_axis({}, {})", ptr->serialize(), axis);

**Review by archibate — COMMENTED** [review]
LGTMig. Let's fix the print(x.shape) issue iapr :)

**Pull request closed without merging**